File: src/backend/app/services/chat/evidence_bundle.py
from __future__ import annotations

from typing import TypedDict

SECTION_KEYS = {
    "changed files": "changedFiles",
    "validations run": "validationsRun",
    "pass/fail evidence": "passFailEvidence",
    "unresolved risks": "unresolvedRisks",
    "completion rationale": "completionRationale",
}


class EvidenceBundleEvent(TypedDict):
    type: str
    schemaVersion: int
    changedFiles: list[str]
    validationsRun: list[str]
    passFailEvidence: list[str]
    unresolvedRisks: list[str]
    completionRationale: list[str]


def _clean_heading(value: str) -> str:
    return value.lstrip("#*- >").rstrip(":* ").strip().lower()


def _normalize_item(line: str) -> str:
    return line.lstrip("*-0123456789. ").strip()


def extract_evidence_bundle_event(text: str) -> EvidenceBundleEvent | None:
    sections: dict[str, list[str]] = {}
    current: str | None = None

    for raw_line in text.splitlines():
        line = raw_line.strip()
        if not line:
            continue
        section_key = SECTION_KEYS.get(_clean_heading(line))
        if section_key is not None:
            current = section_key
            sections.setdefault(section_key, [])
            continue
        if current is None:
            continue
        sections.setdefault(current, []).append(_normalize_item(line))

    if len(sections) < 3:
        return None

    return {
        "type": "evidence.bundle",
        "schemaVersion": 1,
        "changedFiles": sections.get("changedFiles", []),
        "validationsRun": sections.get("validationsRun", []),
        "passFailEvidence": sections.get("passFailEvidence", []),
        "unresolvedRisks": sections.get("unresolvedRisks", []),
        "completionRationale": sections.get("completionRationale", []),
    }

File: src/frontend/src/components/assistant-ui/evidence-bundle.ts
export const EVIDENCE_BUNDLE_SECTIONS = [
  "Changed Files",
  "Validations Run",
  "Pass/Fail Evidence",
  "Unresolved Risks",
  "Completion Rationale",
] as const;

export type EvidenceBundleSectionName = (typeof EVIDENCE_BUNDLE_SECTIONS)[number];

export type EvidenceBundlePayload = {
  schemaVersion?: number;
  changedFiles: string[];
  validationsRun: string[];
  passFailEvidence: string[];
  unresolvedRisks: string[];
  completionRationale: string[];
};

const SECTION_MAP = new Map<string, EvidenceBundleSectionName>(
  EVIDENCE_BUNDLE_SECTIONS.map((label) => [label.toLowerCase(), label]),
);

function cleanHeading(value: string) {
  return value.replace(/^[#*\-\s>]+/, "").replace(/[:*\s]+$/, "").trim().toLowerCase();
}

function normalizeItem(line: string) {
  return line.replace(/^[*\-\d.\s]+/, "").trim();
}

export function parseEvidenceBundleText(text: string): EvidenceBundlePayload | null {
  const sections = new Map<EvidenceBundleSectionName, string[]>();
  let current: EvidenceBundleSectionName | null = null;

  for (const rawLine of text.split("\n")) {
    const line = rawLine.trim();
    if (!line) continue;
    const heading = SECTION_MAP.get(cleanHeading(line));
    if (heading) {
      current = heading;
      if (!sections.has(heading)) sections.set(heading, []);
      continue;
    }
    if (!current) continue;
    sections.get(current)?.push(normalizeItem(line));
  }

  if (sections.size < 3) return null;

  return {
    changedFiles: sections.get("Changed Files") ?? [],
    validationsRun: sections.get("Validations Run") ?? [],
    passFailEvidence: sections.get("Pass/Fail Evidence") ?? [],
    unresolvedRisks: sections.get("Unresolved Risks") ?? [],
    completionRationale: sections.get("Completion Rationale") ?? [],
  };
}
